### PR TITLE
[#89] Reuse existing megaparsec parsers

### LINKED ISSUE #89: Reuse existing megaparsec parsers
# Clarification and motivation

The Web API reimplements some parsing logic to parse things like `Filter` (see the `instance FromHttpApiData Filter`).

We should reuse the same parsers we used in the CLI modules.

Additionally, the logic to parse a `QualifiedPath` is duplicated in `FromHttpApiData (QualifiedPath path)` and in `CLI.Parser.readQualifiedEntryPath`. We should try to have a single `mkQualifiedPath` function that is reused in the `FromHttpApiData` instance and in the CLI parser.

# Acceptance criteria

* The parsing logic is not duplicated.

### PR DESCRIPTION
**Pull request by Kariel-Myrr**
## Problem:
We had a lot of repeating code in `FromHttpApiData` instances with `CLI.Parse`.

## Solution:
That was fixed.
Basically a lot of low parsing logic moved from `CLI.Parse` to `CLI.Types` and `Coffer.Path`.
It was necessary to avoid cyclic dependencies.

## Related issue(s)

- Resolves #89 

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: lib/Coffer/Path.hs [hunk 1]
```diff
@@ -221,16 +224,31 @@ instance (Buildable path) => Buildable (QualifiedPath path) where
...
+  parseUrlPiece = mkQualifiedPath
+    (\_ -> "Invalid qualified path format. Expected [BACKENDNAME#]PATH")
+    parseUrlPiece
+
+mkQualifiedPath
+  :: (Text -> Text)
+  -> (Text -> Either Text path)
+  -> (Text -> Either Text (QualifiedPath path))
+mkQualifiedPath errorMessage mkPath text = case T.splitOn "#" text of
+      [backendNameStr, pathStr] -> do
```
Comment (dcastro) [inline lib/Coffer/Path.hs:236]: Excessive indentation here, let's keep it to 2 spaces

File: lib/Coffer/Path.hs [hunk 2]
```diff
@@ -221,16 +224,31 @@ instance (Buildable path) => Buildable (QualifiedPath path) where
...
+        backendName <- readBackendName' backendNameStr
```
Comment (dcastro) [inline lib/Coffer/Path.hs:237]: We can just use `newBackendName` directly

```suggestion
        backendName <- newBackendName backendNameStr
```

File: lib/Coffer/Path.hs [hunk 3]
```diff
@@ -221,16 +224,31 @@ instance (Buildable path) => Buildable (QualifiedPath path) where
...
+        path <- mkPath pathStr
         pure $ QualifiedPath (Just backendName) path
-    | otherwise =
-        Left "Invalid qualified path format. Expected [BACKENDNAME#]PATH"
+      [pathStr] -> do
+        path <- mkPath pathStr
+        pure $ QualifiedPath Nothing path
+      _ -> Left $ errorMessage text
```
Comment (dcastro) [inline lib/Coffer/Path.hs:243]: Instead of taking in an `errorMessage :: Text -> Text` parameter, let's just return an error message explaining what went wrong, like what the other `mk*` constructors do.

So, what went wrong in the `_` branch? Either `T.splitOn "#"` returned a list with 0 elements or a list with >2 elements.

I don't think `splitOn` ever returns an empty list, at least according to my experiments, so we can rule that out. That means it must have returned a list with >2 elements. Which means the string contained more than one '#' characters, so let's return an error message that says that.

File: lib/CLI/Parser.hs [hunk 4]
```diff
@@ -318,9 +309,9 @@ tagOptions =
 -- Common
 ----------------------------------------------------------------------------
 
-readPath' :: Text -> Either String Path
+readPath' :: Text -> Either Text Path
```
Comment (dcastro) [inline lib/CLI/Parser.hs:312]: Oh, a bit unrelated, but can you please delete `_readEntryPath` while you're here? It's not used anywhere

Comment (dcastro) [inline lib/CLI/Parser.hs:312]: We should do the same to `readEntryPath'` for consistency, change `Either String` to `Either Text`

File: lib/Coffer/Path.hs (hunk 2 above)
Comment (Kariel-Myrr) [reply lib/Coffer/Path.hs]: Done

File: lib/Coffer/Path.hs (hunk 3 above)
Comment (Kariel-Myrr) [reply lib/Coffer/Path.hs]: Done

File: lib/CLI/Parser.hs (hunk 4 above)
Comment (Kariel-Myrr) [reply lib/CLI/Parser.hs:312]: Done

Comment (Kariel-Myrr) [reply lib/CLI/Parser.hs:312]: Done

**Review by dcastro — APPROVED** [review]

File: tests/golden/common/common.bats [hunk 5]
```diff
@@ -43,7 +43,6 @@ EOF
```
Comment (DK318) [inline tests/golden/common/common.bats:43]: Imo it would be better if we split fixing this test case in another commit.

**Review by DK318 — APPROVED** [review]
LGTM overall

File: tests/golden/common/common.bats (hunk 5 above)
Comment (Kariel-Myrr) [reply tests/golden/common/common.bats:43]: I could, but it would mean on one stage our test's won't  pass. Is it acceptable? And this changes directly connected to code changes
